### ISSUE TITLE
Consider support GCP Storage Buckets as storage backend (similar to S3 Storage)

### ISSUE THREAD

**Issue by SimSama** [ClickHouse/ClickHouse#39139]
Currently it appears that AWS S3 bucket object storage can be used as a storage backend for Clickhouse, which is great. 
Are there plans in the works for Google Cloud (GCP) Storage Buckets (their version of object storage) to be used as a storage backend?

Related: I know that GCP Storage Buckets have S3 API emulation, but worried that 1:1 compatibility may not work. Has anyone tried this in production? e.g. Using Clickhouse S3 plugin facing GCP storage bucket, which is emulating an S3 bucket?

Use Case 1)
Dual write data to hot/warm local disks, and archive a copy to object storage in cloud for long term retention. In disaster recovery scenario, can load data from object storage.

Use Case 2)
Use object storage as sole data backend instead of local disk (Not sure if this is done today for S3)

Let me know what you think.

Thanks!

**Comment by UnamedRus** [comment]
> Has anyone tried this in production? 
>  Use object storage as sole data backend instead of local disk (Not sure if this is done today for S3)

No, because there was some issues with DeleteObjects (gcp doesn't support it) and ListObjectsV2 (didn't supported in before, but now it's working)

Fix for DeleteObjects was implemented recently:
https://github.com/ClickHouse/ClickHouse/pull/37882
https://github.com/ClickHouse/ClickHouse/pull/37659

One of latest pieces: MultiPart upload, it's being done slightly different in gcp

https://kb.altinity.com/altinity-kb-integrations/altinity-kb-google-s3-gcs/

**Comment by SimSama** [comment]
Is the issue with MultiPart upload still the case? Google's release notes state it went GA in July 12 , 2021, and that altinity doc was lost modified about the same time frame a few months later?

https://cloud.google.com/storage/docs/release-notes#July_12_2021
https://cloud.google.com/storage/docs/multipart-uploads

My original feature request however was, if Clickhouse would consider supporting GCP storage buckets natively w/o using the S3 emulation XML API. That would eliminate this issue with 1:1 implementation matching as the S3 API evolves. 

Thanks,

**Comment by alexey-milovidov** [comment]
It works!
